File: api/explain.py
import sqlite3
import json
from http.server import BaseHTTPRequestHandler
import os

def get_scheme_from_db(scheme_name):
    """Query SQLite database for scheme details"""
    try:
        db_path = os.path.join(os.path.dirname(__file__), 'schemes.db')
        conn = sqlite3.connect(db_path)
        conn.row_factory = sqlite3.Row
        cursor = conn.cursor()
        
        cursor.execute("""
            SELECT * FROM schemes 
            WHERE LOWER(name) LIKE LOWER(?)
            LIMIT 1
        """, (f'%{scheme_name}%',))
        
        row = cursor.fetchone()
        conn.close()
        
        if row:
            return dict(row)
        return None
    except Exception as e:
        print(f"Database error: {e}")
        return None

def get_scheme_from_json(scheme_name):
    """Fallback to JSON file"""
    try:
        json_path = os.path.join(os.path.dirname(__file__), 'schemes.json')
        with open(json_path, 'r') as f:
            schemes = json.load(f)
        
        for scheme in schemes:
            if scheme_name.lower() in scheme['name'].lower():
                return scheme
        return None
    except Exception as e:
        print(f"JSON error: {e}")
        return None

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        if self.path == '/api/explain':
            content_length = int(self.headers['Content-Length'])
            body = self.rfile.read(content_length)
            data = json.loads(body.decode('utf-8'))
            
            scheme_name = data.get('schemeName', '')
            
            if not scheme_name:
                self.send_response(400)
                self.send_header('Content-type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Scheme name required'}).encode())
                return
            
            scheme = get_scheme_from_db(scheme_name) or get_scheme_from_json(scheme_name)
            
            if scheme:
                response = {
                    'summary': scheme.get('description', 'No description available'),
                    'eligibility': scheme.get('eligibility', 'Eligibility criteria not specified'),
                    'benefits': scheme.get('benefits', 'Benefits information not available'),
                    'process': scheme.get('steps', 'Application process not specified')
                }
            else:
                response = {
                    'summary': f'{scheme_name} is a government initiative. Detailed information is being updated.',
                    'eligibility': 'Please check the official government portal for eligibility criteria.',
                    'benefits': 'Benefits vary based on the scheme. Visit the official website for details.',
                    'process': 'Apply through the official government portal or nearest Common Service Center.'
                }
            
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps(response).encode())
        else:
            self.send_response(404)
            self.end_headers()
    
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

File: api/explain.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const schemesPath = path.join(__dirname, 'schemes_cleaned.json');
let schemes = [];

try {
  const data = fs.readFileSync(schemesPath, 'utf8');
  schemes = JSON.parse(data);
} catch (error) {
  console.error('Error loading schemes:', error);
}

function findScheme(schemeName) {
  const searchTerm = schemeName.toLowerCase().trim();
  return schemes.find(scheme => 
    scheme.scheme_name.toLowerCase().includes(searchTerm) ||
    scheme.scheme_name.toLowerCase().replace(/\s+/g, '').includes(searchTerm.replace(/\s+/g, ''))
  );
}

function formatEligibility(eligibility) {
  if (typeof eligibility === 'string') return eligibility;
  if (typeof eligibility === 'object') {
    const parts = [];
    if (eligibility.age) parts.push(`Age: ${eligibility.age}`);
    if (eligibility.income) parts.push(`Income: ${eligibility.income}`);
    if (eligibility.gender) parts.push(`Gender: ${eligibility.gender}`);
    if (eligibility.category) parts.push(`Category: ${eligibility.category}`);
    if (eligibility.occupation) parts.push(`Occupation: ${eligibility.occupation}`);
    if (eligibility.state) parts.push(`State: ${eligibility.state}`);
    return parts.join('. ') + '.';
  }
  return 'Eligibility criteria not specified';
}

export default async (req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }

  if (req.method !== 'POST') {
    res.status(405).json({ error: 'Method not allowed' });
    return;
  }

  const { schemeName, simplify } = req.body;

  if (!schemeName) {
    res.status(400).json({ error: 'Scheme name is required' });
    return;
  }

  const scheme = findScheme(schemeName);

  if (scheme) {
    let summary = scheme.description || 'No description available';
    let eligibility = formatEligibility(scheme.eligibility);
    let benefits = scheme.benefits || 'Benefits information not available';
    let process = `To apply for ${scheme.scheme_name}: Visit the official portal at ${scheme.source_url || 'the government website'}. Required documents: ${scheme.documents_required?.join(', ') || 'Check official website'}. Submit your application online or at the nearest Common Service Center.`;

    // Simplify if requested
    if (simplify) {
      summary = summary.split('.')[0] + '.';
      eligibility = eligibility.split('.').slice(0, 2).join('. ') + '.';
      benefits = benefits.split('.')[0] + '.';
      process = process.split('.').slice(0, 2).join('. ') + '.';
    }

    res.status(200).json({
      summary,
      eligibility,
      benefits,
      process,
      scheme_name: scheme.scheme_name,
      source_url: scheme.source_url
    });
  } else {
    res.status(404).json({
      error: 'Scheme not found',
      summary: `${schemeName} scheme not found in our database.`,
      eligibility: 'Please check the scheme name and try again.',
      benefits: 'Visit the official government portal for details.',
      process: 'Search for the correct scheme name or browse all schemes.'
    });
  }
};
